File: internal/billing/credit_cache.go
package billing

import (
	"context"
	_ "embed"
	"fmt"

	"github.com/redis/go-redis/v9"
)

//go:embed lua/credit_cache.lua
var creditCacheLuaScript string
var creditCacheScript *redis.Script

func init() {
	creditCacheScript = redis.NewScript(creditCacheLuaScript)
}

// CreditBalanceCache 对用户某个余额池的 Redis 缓存按 delta 原子增减（delta 可负，用于 debit）。
//
// 仅在缓存 key 已存在时生效——避免用绝对值覆盖请求侧 DECRBY 的实时扣减导致丢扣/超额消费。
// key 不存在时跳过，由下次请求 miss 从（已在同一笔 DB 事务中更新过的）DB 懒加载，最终一致。
func CreditBalanceCache(ctx context.Context, userID int32, balanceType string, delta int64) error {
	if RedisClient == nil {
		return fmt.Errorf("redis client is not initialized")
	}

	var key string
	switch balanceType {
	case "grant":
		key = fmt.Sprintf("grant_balance:%d", userID)
	case "cash":
		key = fmt.Sprintf("cash_balance:%d", userID)
	default:
		return fmt.Errorf("unsupported balance type: %s", balanceType)
	}

	return creditCacheScript.Run(ctx, RedisClient, []string{key}, delta).Err()
}

File: internal/billing/lua/credit_cache.lua
-- 充值/扣账对余额缓存的原子相对增减。
-- KEYS[1]: 余额 key (grant_balance:uid 或 cash_balance:uid)
-- ARGV[1]: 变动量 delta（可正可负；credit 为正，debit 为负）
--
-- 仅当 key 已存在时才 INCRBY：缓存-aside 语义下，key 存在表示它正反映实时余额
-- （请求侧用 DECRBY 扣减），此时按 delta 相对增减可与并发扣减正确叠加，避免用绝对
-- SET 覆盖掉在途扣减。key 不存在则跳过，由下次请求 miss 时从已更新的 DB 懒加载。
if redis.call("EXISTS", KEYS[1]) == 1 then
    return redis.call("INCRBY", KEYS[1], ARGV[1])
end
return -1
